File: app_video_edit/src/main/java/com/gxz/example/videoedit/VideoEditActivity.java
package com.gxz.example.videoedit;

import android.animation.ValueAnimator;
import android.arch.lifecycle.Observer;
import android.media.MediaPlayer;
import android.os.Bundle;
import android.os.Environment;
import android.os.Handler;
import android.os.Looper;
import android.os.Message;
import android.support.annotation.Nullable;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.View;
import android.view.animation.LinearInterpolator;
import android.widget.FrameLayout;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;
import android.widget.Toast;
import android.widget.VideoView;

import java.io.File;
import java.lang.ref.WeakReference;


public class VideoEditActivity extends AppCompatActivity {
    private static final String TAG = VideoEditActivity.class.getSimpleName();
    private LinearLayout seekBarLayout;
    private RangeSeekBar seekBar;
    private VideoView mVideoView;
    private RecyclerView mRecyclerView;
    private ImageView positionIcon;
    private TextView textViewInfo;

    private VideoEditAdapter videoEditAdapter;
    private String path;
    private long leftProgress;
    private long rightProgress;

    private Handler UIHandler = new Handler(Looper.getMainLooper());
    private VideoCutViewModel videoCutViewModel;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_video_edit);
        initData();
        initView();

        initPlay();
        initEditVideo();
    }

    private void initData() {
        path = Environment.getExternalStorageDirectory() + "/2.mp4";
        //for video check
        if (!new File(path).exists()) {
            Toast.makeText(this, "视频文件不存在", Toast.LENGTH_LONG).show();
            finish();
        }
        int mMaxWidth = UIUtil.getScreenWidth(this) - UIUtil.dip2px(this, 70);
        videoCutViewModel = new VideoCutViewModel(this, path, mMaxWidth);
        videoCutViewModel.getVideoPauseEvent().observe(this, new Observer<Void>() {
            @Override
            public void onChanged(@Nullable Void aVoid) {
                videoPause();
            }
        });

        videoCutViewModel.getVideoStartEvent().observe(this, new Observer<Void>() {
            @Override
            public void onChanged(@Nullable Void aVoid) {
                videoStart();
            }
        });

        videoCutViewModel.getVideoSeekEvent().observe(this, new Observer<Integer>() {
            @Override
            public void onChanged(@Nullable Integer integer) {
                mVideoView.seekTo(integer);
            }
        });

        videoCutViewModel.getStartProgress().observe(this, new Observer<Long>() {
            @Override
            public void onChanged(@Nullable Long aLong) {
                leftProgress = aLong;
                showPosInfo(leftProgress, rightProgress);
            }
        });

        videoCutViewModel.getEndProgress().observe(this, new Observer<Long>() {
            @Override
            public void onChanged(@Nullable Long aLong) {
                rightProgress = aLong;
                showPosInfo(leftProgress, rightProgress);
            }
        });

        videoCutViewModel.getThumbInfo().observe(this, new Observer<VideoEditInfo>() {
            @Override
            public void onChanged(@Nullable VideoEditInfo videoEditInfo) {
                videoEditAdapter.addItemVideoInfo(videoEditInfo);
            }
        });
    }

    private void initView() {
        textViewInfo = (TextView) findViewById(R.id.tv_debug);
        seekBarLayout = (LinearLayout) findViewById(R.id.id_seekBarLayout);
        mVideoView = (VideoView) findViewById(R.id.uVideoView);
        positionIcon = (ImageView) findViewById(R.id.positionIcon);
        mRecyclerView = (RecyclerView) findViewById(R.id.id_rv_id);
        mRecyclerView.setLayoutManager(new LinearLayoutManager(this, LinearLayoutManager.HORIZONTAL, false));
        videoEditAdapter = new VideoEditAdapter(this,
                (UIUtil.getScreenWidth(this) - UIUtil.dip2px(this, 70)) / VideoCutViewModel.MAX_COUNT_RANGE);
        mRecyclerView.setAdapter(videoEditAdapter);
        mRecyclerView.addOnScrollListener(mOnScrollListener);
    }


    private void initEditVideo() {
        mRecyclerView.addItemDecoration(new EditSpacingItemDecoration(UIUtil.dip2px(this, 35), videoCutViewModel.getThumbPicNum()));

        seekBar = new RangeSeekBar(this);
        seekBar.setMinDistanceRatio(1.0 * VideoCutViewModel.MIN_CUT_DURATION / VideoCutViewModel.MAX_CUT_DURATION);
        seekBar.setOnRangeSeekBarChangeListener(mOnRangeSeekBarChangeListener);
        seekBarLayout.addView(seekBar);

        videoCutViewModel.start();
    }


    private void initPlay() {
        mVideoView.setVideoPath(path);
        mVideoView.setOnPreparedListener(new MediaPlayer.OnPreparedListener() {
            @Override
            public void onPrepared(MediaPlayer mp) {
                mp.setOnSeekCompleteListener(new MediaPlayer.OnSeekCompleteListener() {
                    @Override
                    public void onSeekComplete(MediaPlayer mp) {
                        videoCutViewModel.onSeekComplete();
                    }
                });
            }
        });
    }


    private final RecyclerView.OnScrollListener mOnScrollListener = new RecyclerView.OnScrollListener() {
        @Override
        public void onScrollStateChanged(RecyclerView recyclerView, int newState) {
            super.onScrollStateChanged(recyclerView, newState);
            videoCutViewModel.onScrollStateChanged(newState);
        }

        @Override
        public void onScrolled(RecyclerView recyclerView, int dx, int dy) {
            super.onScrolled(recyclerView, dx, dy);
            int scrollX = getScrollXDistance();
            videoCutViewModel.onScrolled(scrollX, dx, dy);
        }
    };

    /**
     * 水平滑动了多少px
     */
    private int getScrollXDistance() {
        LinearLayoutManager layoutManager = (LinearLayoutManager) mRecyclerView.getLayoutManager();
        int position = layoutManager.findFirstVisibleItemPosition();
        View firstVisibleChildView = layoutManager.findViewByPosition(position);
        int itemWidth = firstVisibleChildView.getWidth();
        return (position) * itemWidth - firstVisibleChildView.getLeft();
    }

    private ValueAnimator animator;

    private void anim() {
        Log.e(TAG, "anim start pos: " + mVideoView.getCurrentPosition());
        if (positionIcon.getVisibility() == View.GONE) {
            positionIcon.setVisibility(View.VISIBLE);
        }
        final FrameLayout.LayoutParams params = (FrameLayout.LayoutParams) positionIcon.getLayoutParams();

        int start = (int) (UIUtil.dip2px(this, 35) + seekBar.getLeftOffset());
        int end = (int) (UIUtil.dip2px(this, 35) + seekBar.getRightOffset());

        animator = ValueAnimator
                .ofInt(start, end)
                .setDuration(videoCutViewModel.getVideoDuration());
        animator.setInterpolator(new LinearInterpolator());
        final int finalEnd = end;
        animator.addUpdateListener(new ValueAnimator.AnimatorUpdateListener() {
            @Override
            public void onAnimationUpdate(ValueAnimator animation) {
                params.leftMargin = (int) animation.getAnimatedValue();
                positionIcon.setLayoutParams(params);

                if(params.leftMargin == finalEnd){
                    Log.e(TAG, "anim end pos: " + mVideoView.getCurrentPosition());
                }
            }
        });
        animator.start();
    }

    private final RangeSeekBar.OnRangeSeekBarChangeListener mOnRangeSeekBarChangeListener = new RangeSeekBar.OnRangeSeekBarChangeListener() {
        @Override
        public void onRangeSeekBarValuesChanged(int action, boolean isLeftHandleMoving) {
            videoCutViewModel.onRangeSeekBarValuesChanged(isLeftHandleMoving ? seekBar.getLeftOffset() : seekBar.getRightOffset(),
                    seekBar.getWidth(), isLeftHandleMoving, action);
        }
    };


    private void videoStart() {
        UIHandler.post(new Runnable() {
            @Override
            public void run() {
                Log.d(TAG, "----videoStart----->>>>>>>");
                mVideoView.start();
                positionIcon.clearAnimation();
                if (animator != null && animator.isRunning()) {
                    animator.cancel();
                }
                anim();
                handler.removeCallbacks(ReplayTask);
                handler.postDelayed(ReplayTask, videoCutViewModel.getVideoDuration());
            }
        });
    }

    private boolean replayVideoIfNeed() {
        boolean needReplay = mVideoView.getCurrentPosition() >= rightProgress;
        if (needReplay) {
            Log.e(TAG, "check replay true pos: " + mVideoView.getCurrentPosition());
            mVideoView.pause(); //有必要，不然再次播放时，seek的指针位置有变化
            mVideoView.seekTo((int) leftProgress);
            positionIcon.clearAnimation();
            if (animator != null && animator.isRunning()) {
                animator.cancel();
            }
        }else {
            Log.e(TAG, "check replay false pos: " + mVideoView.getCurrentPosition());
        }

        return needReplay;
    }

    private void videoPause() {
        if(mVideoView == null || !mVideoView.isPlaying()) return;
        if (mVideoView != null && mVideoView.isPlaying()) {
            mVideoView.pause();
            handler.removeCallbacks(ReplayTask);
        }
        Log.d(TAG, "----videoPause----->>>>>>>");
        if (positionIcon.getVisibility() == View.VISIBLE) {
            positionIcon.setVisibility(View.GONE);
        }
        positionIcon.clearAnimation();
        if (animator != null && animator.isRunning()) {
            animator.cancel();
        }
    }


    @Override
    protected void onResume() {
        super.onResume();
        if (mVideoView != null) {
            mVideoView.seekTo((int) leftProgress);
        }
    }

    @Override
    protected void onPause() {
        super.onPause();
        if (mVideoView != null && mVideoView.isPlaying()) {
            videoPause();
        }
    }

    private Handler handler = new Handler();
    private Runnable ReplayTask = new Runnable() {

        @Override
        public void run() {
            boolean needReplay = replayVideoIfNeed();
            if(!needReplay) {
                long nextCheckTime = rightProgress - mVideoView.getCurrentPosition(); //下次检测是否播放完成的时间点
                handler.postDelayed(ReplayTask, nextCheckTime);
            }
        }
    };

    @Override
    protected void onDestroy() {
        super.onDestroy();
        if (animator != null) {
            animator.cancel();
        }
        if (mVideoView != null) {
            mVideoView.stopPlayback();
        }
        mRecyclerView.removeOnScrollListener(mOnScrollListener);
        handler.removeCallbacksAndMessages(null);
        videoCutViewModel.end();
    }

    private void showPosInfo(long leftPos, long rightProgress){
        textViewInfo.setText(leftPos + " : " + rightProgress);
    }
}

File: app_video_edit/src/main/java/com/gxz/example/videoedit/VideoEditActivity.kt
package com.gxz.example.videoedit

import android.animation.ValueAnimator
import androidx.lifecycle.Observer
import android.os.Bundle
import android.os.Environment
import android.os.Handler
import android.os.Looper
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import android.util.Log
import android.view.View
import android.view.animation.LinearInterpolator
import android.widget.FrameLayout
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView
import android.widget.Toast
import android.widget.VideoView

import java.io.File


class VideoEditActivity : AppCompatActivity() {
    private var seekBarLayout: LinearLayout? = null
    private lateinit var seekBar: VideoSeekBar
    private lateinit var mVideoView: VideoView
    private var mRecyclerView: androidx.recyclerview.widget.RecyclerView? = null
    private var positionIcon: ImageView? = null
    private var textViewInfo: TextView? = null

    private var videoEditAdapter: VideoEditAdapter? = null
    private var path: String? = null
    private var leftProgress: Long = 0
    private var rightProgress: Long = 0

    private val UIHandler = Handler(Looper.getMainLooper())
    private var videoCutViewModel: VideoCutViewModel? = null

    private var animator: ValueAnimator? = null
    private val handler = Handler()
    private val replayTask = object : Runnable {

        override fun run() {
            val needReplay = replayVideoIfNeed()
            if (!needReplay) {
                val nextCheckTime = rightProgress - mVideoView.currentPosition //下次检测是否播放完成的时间点
                handler.postDelayed(this, nextCheckTime)
            }
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_video_edit)
        initData()
        initView()

        initPlay()
        initEditVideo()
    }

    private fun initData() {
        path = Environment.getExternalStorageDirectory().toString() + "/2.mp4"
        //for video check
        if(!File(path).exists()){
            Toast.makeText(this, "视频文件不存在", Toast.LENGTH_LONG).show()
            finish()
        }
        val mMaxWidth = UIUtil.getScreenWidth(this) - UIUtil.dip2px(this, 70)
        videoCutViewModel = VideoCutViewModel(this, path, mMaxWidth)
        videoCutViewModel?.videoPauseEvent?.observe(this, Observer { videoPause() })

        videoCutViewModel?.videoStartEvent?.observe(this, Observer { videoStart() })

        videoCutViewModel?.videoSeekEvent?.observe(this, Observer { integer ->
            if (integer == 0) {
                videoStart()
            } else {
                mVideoView.seekTo(integer?:0)
            }
        })

        videoCutViewModel?.startProgress?.observe(this, Observer { aLong ->
            leftProgress = aLong ?: 0
            showPosInfo(leftProgress, rightProgress)
        })

        videoCutViewModel?.endProgress?.observe(this, Observer { aLong ->
            rightProgress = aLong ?: 0
            showPosInfo(leftProgress, rightProgress)
        })

        videoCutViewModel?.thumbInfo?.observe(this, Observer { videoEditInfo -> videoEditAdapter?.addItemVideoInfo(videoEditInfo) })

        videoCutViewModel?.let {
            lifecycle.addObserver(it)
        }
    }


    private val mOnScrollListener = object : androidx.recyclerview.widget.RecyclerView.OnScrollListener() {
        override fun onScrollStateChanged(recyclerView: androidx.recyclerview.widget.RecyclerView?, newState: Int) {
            super.onScrollStateChanged(recyclerView, newState)
            videoCutViewModel?.onScrollStateChanged(newState)
        }

        override fun onScrolled(recyclerView: androidx.recyclerview.widget.RecyclerView?, dx: Int, dy: Int) {
            super.onScrolled(recyclerView, dx, dy)
            val scrollX = scrollXDistance + UIUtil.dip2px(this@VideoEditActivity, 35)
            videoCutViewModel?.onScrolled(scrollX, dx, dy)
        }
    }

    private val mOnRangeSeekBarChangeListener = VideoSeekBar.OnRangeSeekBarChangeListener { action, isLeftHandleMoving ->
        var offset = if (isLeftHandleMoving) seekBar.leftOffset else seekBar.rightOffset
        videoCutViewModel?.onRangeSeekBarValuesChanged(offset, seekBar.width.toFloat(), isLeftHandleMoving, action)
    }


    private fun initView() {
        textViewInfo = findViewById<View>(R.id.tv_debug) as TextView
        seekBarLayout = findViewById<View>(R.id.id_seekBarLayout) as LinearLayout
        mVideoView = findViewById<View>(R.id.uVideoView) as VideoView
        positionIcon = findViewById<View>(R.id.positionIcon) as ImageView
        mRecyclerView = findViewById<View>(R.id.id_rv_id) as androidx.recyclerview.widget.RecyclerView
        mRecyclerView?.layoutManager = androidx.recyclerview.widget.LinearLayoutManager(this, androidx.recyclerview.widget.LinearLayoutManager.HORIZONTAL, false)
        videoEditAdapter = VideoEditAdapter(this,
                (UIUtil.getScreenWidth(this) - UIUtil.dip2px(this, 70)) / VideoCutViewModel.MAX_COUNT_RANGE)
        mRecyclerView?.adapter = videoEditAdapter
        mRecyclerView?.addOnScrollListener(mOnScrollListener)
    }


    private fun initEditVideo() {
        mRecyclerView?.addItemDecoration(EditSpacingItemDecoration(UIUtil.dip2px(this, 35), videoCutViewModel?.getThumbPicNum() ?: 0))

        seekBar = VideoSeekBar(this)
        seekBar.setMinDistanceRatio(1.0 * VideoCutViewModel.MIN_CUT_DURATION / VideoCutViewModel.MAX_CUT_DURATION)
        seekBar.setOnRangeSeekBarChangeListener(mOnRangeSeekBarChangeListener)
        seekBarLayout?.addView(seekBar)
    }

    private fun initPlay() {
        mVideoView.setVideoPath(path)
        mVideoView.setOnPreparedListener { mp -> mp.setOnSeekCompleteListener { videoCutViewModel?.onSeekComplete() } }
    }

    private fun startAnim() {
        Log.e(TAG, "startAnim start pos: " + mVideoView.currentPosition)
        if (positionIcon?.visibility == View.GONE) {
            positionIcon?.visibility = View.VISIBLE
        }
        val params = positionIcon?.layoutParams as FrameLayout.LayoutParams

        val start = (UIUtil.dip2px(this, 35) + seekBar.leftOffset).toInt()
        val end = (UIUtil.dip2px(this, 35) + seekBar.rightOffset).toInt()

        videoCutViewModel?.let {
            animator = ValueAnimator
                    .ofInt(start, end)
                    .setDuration(it.getVideoDuration())
            animator?.interpolator = LinearInterpolator()
            animator?.addUpdateListener { animation ->
                params.leftMargin = animation.animatedValue as Int
                positionIcon?.layoutParams = params

                if (params.leftMargin == end) {
                    Log.e(TAG, "startAnim end pos: " + mVideoView.currentPosition)
                }
            }
            animator?.start()
        }
    }

    private fun stopAnim() {
        if (positionIcon?.visibility == View.VISIBLE) {
            positionIcon?.visibility = View.GONE
        }
        positionIcon?.clearAnimation()
        if (animator?.isRunning == true) {
            animator?.cancel()
        }
    }

    private fun videoStart() {
        videoCutViewModel?.let {
            UIHandler.post {
                Log.d(TAG, "----videoStart----->>>>>>>")
                mVideoView.start()
                stopAnim()
                startAnim()
                handler.removeCallbacks(replayTask)
                handler.postDelayed(replayTask, it.getVideoDuration())
            }
        }
    }

    private fun replayVideoIfNeed(): Boolean {
        val needReplay = mVideoView.currentPosition >= rightProgress
        if (needReplay) {
            Log.e(TAG, "check replay true pos: " + mVideoView.currentPosition)
            mVideoView.pause() //有必要，不然再次播放时，seek的指针位置有变化
            mVideoView.seekTo(leftProgress.toInt())
            stopAnim()
        } else {
            Log.e(TAG, "check replay false pos: " + mVideoView.currentPosition)
        }

        return needReplay
    }

    private fun videoPause() {
        if (mVideoView.isPlaying) {
            mVideoView.pause()
            handler.removeCallbacks(replayTask)
            Log.d(TAG, "----videoPause----->>>>>>>")
            stopAnim()
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        animator?.cancel()
        mVideoView.stopPlayback()
        mRecyclerView?.removeOnScrollListener(mOnScrollListener)
        handler.removeCallbacksAndMessages(null)
        videoCutViewModel?.let {
            it.end()
            lifecycle.removeObserver(it)
        }
    }

    private fun showPosInfo(leftPos: Long, rightProgress: Long) {
        textViewInfo?.text = "$leftPos : $rightProgress"
    }


    private val scrollXDistance: Int
        get() {
            val layoutManager = mRecyclerView?.layoutManager as androidx.recyclerview.widget.LinearLayoutManager
            val position = layoutManager.findFirstVisibleItemPosition()
            val firstVisibleChildView = layoutManager.findViewByPosition(position)
            val itemWidth = firstVisibleChildView.width
            return position * itemWidth - firstVisibleChildView.left
        }

    companion object {
        private val TAG = VideoEditActivity::class.java.simpleName
    }
}
